### DISCUSSION TITLE
Read an app setting in my .Net core application without using DI

### DISCUSSION THREAD

**Discussion by csimone86** [dotnet/aspnetcore#29628] [Q&A]
I created an MVC application using .Net core. Then I created a custom Attribute should read some settings from appsettings.json. Here a working solution:

```
public class MyCustomAttribute : Attribute
{
    public MyCustomAttribute(string key)
        : base()
    {
        IConfiguration conf = new ConfigurationBuilder()
           .SetBasePath(Directory.GetCurrentDirectory())
           .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
           .Build();
        
        var value = conf.GetValue<string>(key);

        ...
    }

    ...
}
```

It works, but I do not think it is the correct solution. I try to explain why.

In the Program.cs, beforse the host is builded, the startup class is instantiated:

```
var builder = Host.CreateDefaultBuilder(args); builder
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.UseStartup<Startup>();

    });
```
If I inject the IConfiguration to the stratup, I can get the appsettings.json values.

I think that from now, I have the appsettings.json and conseguently the configuration in memory.

So it seems really strange to me that in my custom attribute I must read the setting from the file again! The question is: How can I read the in-memory configuration inside my custom attribute? Is my consideration correct?

Thank you.

**Comment by Tratcher** [comment]
Where is the attribute consumed? I would expect the attribute to only hold the input `key` value and you'd handle the config reading later at the consumption point.

**Reply by csimone86** [reply]
Hi Tratcher,
thank you for your answer...  The attribute is used on a property on a ViewModel... and my custom attribute inherit from a ValidationAttribute. I let you see:

```
    public class MyCustomViewModel
    {
         ....
         
        [MyCustomRemote(RemoteUrls.EmailValidationAppSettingKey)]
        public string Email { get; set; }
        ....
    }
```

That is my `MyCustomRemoteAttribute`:

```
    public class MyCustomRemoteAttribute : Microsoft.AspNetCore.Mvc.RemoteAttribute
    {
        private string _url;

        public RemoteAttribute(string key)
            : base()
        {
            //Here I need to read configuration 
            this._url = conf.GetValue<string>(key);
        }
    ....
    }
```



While you have answered my I have found another solution. But I still do not like it! (It's two days I am working on :( )
I let you see my new solution:

In the `startup.cs` now I have:

```
    public class Startup
    {
        private IWebHostEnvironment _environment { get; }
        private IConfiguration _configuration;

        public Startup(IWebHostEnvironment environment, IConfiguration configuration)
        {
            this._environment = environment;
            this._configuration = configuration;
        }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        { 
             AppDomain.CurrentDomain.SetData("conf", this._configuration);
             ....
         }
```

And `MyCustomRemoteAttribute`  ctor has become: 

```
        public RemoteAttribute(string key)
            : base()
        {
            var conf = AppDomain.CurrentDomain.GetData("conf") as IConfiguration;
            this._url = conf.GetValue<string>(key);
        }
```

What I would expect is that .net core do it "automatically" (out-of-the-box?) and I just have to access the correct "Context" (AppContext?).
I am trying doing reverse engineering on .net core source code but (obviously) no success.... 

I hope I have been clear.

Thank you again.

**Reply by Tratcher** [reply]
ASP.NET Core has been intentionally designed to avoid static state, it would severely limit testability and extensibility of the system.

**Reply by csimone86** [reply]
Cannot either use a service locator?

**Reply by Tratcher** [reply]
How so?
